src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                    TimeSpan currentTime = (PreciseTimeSpan.FromStart - this.owner.StartTime).ToTimeSpan();
+                    TimeSpan sleepTime = CeilTimeSpanToMilliseconds(TimeSpan.FromTicks(deadline - currentTime.Ticks));
+
+                    if (sleepTime <= TimeSpan.Zero)
+                    {
+                        if (currentTime.Ticks == long.MinValue)
+                        {
+                            return TimeSpan.FromTicks(-long.MaxValue);
+                        }
+                        else
+                        {
+                            return currentTime;
+                        }
+                    }
+
+                    Task delay = null;
+                    try
+                    {
+                        long sleepTimeMs = sleepTime.Ticks / TimeSpan.TicksPerMillisecond; // we've already rounded so no worries about the remainder > 0 here
+                        Contract.Assert(sleepTimeMs <= int.MaxValue);
+                        delay = Task.Delay((int)sleepTimeMs, this.owner.CancellationToken);
+                        delay.Wait();
+                    }
+                    catch (AggregateException) when (delay != null && delay.IsCanceled)
+                    {
+                        if (Volatile.Read(ref this.owner.workerStateVolatile) == WorkerStateShutdown)
+                        {
+                            return TimeSpan.FromTicks(long.MinValue);
+                        }
+                    }
+                }
+            }
+
+            internal ISet<ITimeout> UnprocessedTimeouts() => this.unprocessedTimeouts;
+        }
+
+        sealed class HashedWheelTimeout : ITimeout
+        {
+            const int StInit = 0;
+            internal const int StCancelled = 1;
+            const int StExpired = 2;
+
+            internal readonly HashedWheelTimer timer;
+            internal readonly TimeSpan Deadline;
+
+            volatile int state = StInit;
+
+            // remainingRounds will be calculated and set by Worker.transferTimeoutsToBuckets() before the
+            // HashedWheelTimeout will be added to the correct HashedWheelBucket.
+            internal long RemainingRounds;
+
+            // This will be used to chain timeouts in HashedWheelTimerBucket via a double-linked-list.
+            // As only the workerThread will act on it there is no need for synchronization / volatile.
+            internal HashedWheelTimeout Next;
+
+            internal HashedWheelTimeout Prev;
+
+            // The bucket to which the timeout was added
+            internal HashedWheelBucket Bucket;
+
+            internal HashedWheelTimeout(HashedWheelTimer timer, ITimerTask task, TimeSpan deadline)
+            {
+                this.timer = timer;
+                this.Task = task;
+                this.Deadline = deadline;
+            }
+
+            public ITimer Timer => this.timer;
+
+            public ITimerTask Task { get; }
+
+            public bool Cancel()
```

Comment:
if the HashedWheelTimer is shutdown and then later someone calls Cancel(), then this will just leak an entry on the IQueue object on the timer class, right? should we not enqueue if the timer is shutdown?